internal/controller/buildpipeline/buildpipeline_adapter.go

```diff
@@ -650,3 +691,112 @@ func generateDetails(buildPLR *tektonv1.PipelineRun, integrationTestStatus intgt
+	prGroupHash, prGroup := gitops.GetPRGroup(pipelineRun)
+	componentsFromPipelineRun, err := a.getComponentFromLatestFlightBuildPLR(prGroup, prGroupHash)
+	if err != nil {
+		a.logger.Error(err, fmt.Sprintf("failed to get component from build pipelineruns for pr group %s", prGroup))
+		return false, err
+	}
+
+	componentsFromSnapshot, err := a.loader.GetComponentsFromSnapshotForPRGroup(a.context, a.client, pipelineRun.Namespace, prGroup, prGroupHash)
+	if err != nil {
+		a.logger.Error(err, fmt.Sprintf("failed to get component from component snapshot for pr group %s", prGroup))
+		return false, err
+	}
+
+	joinedComponents := gitops.MergeAndDeduplicate(componentsFromPipelineRun, componentsFromSnapshot)
+	if len(joinedComponents) < 2 {
+		a.logger.Info(fmt.Sprintf("The number %s of components affected by this PR group %s is less than 2, skipping group snapshot creation", joinedComponents, prGroup))
+		return false, nil
+	}
+
+	applicationComponents, err := a.loader.GetAllApplicationComponents(a.context, a.client, a.application)
+	if err != nil {
+		return false, err
+	}
+
+	// check if there is opened MR/PR for the components
+	var componentsWithOpenPRMR []string
+	for _, applicationComponent := range *applicationComponents {
+		applicationComponent := applicationComponent // G601
+
+		if slices.Contains(componentsFromPipelineRun, applicationComponent.Name) {
+			a.logger.Info(fmt.Sprintf("There is in progress build pipelineRun for component %s/%s, won't check its component snapshot's pull/merge request", a.pipelineRun.Namespace, applicationComponent.Name))
+			componentsWithOpenPRMR = append(componentsWithOpenPRMR, applicationComponent.Name)
+			continue
+		}
+
+		if slices.Contains(componentsFromSnapshot, applicationComponent.Name) {
+			a.logger.Info(fmt.Sprintf("There is component snapshot for component %s/%s, will check its component snapshot MR/PR status", a.pipelineRun.Namespace, applicationComponent.Name))
+			var foundSnapshotWithOpenedPR *applicationapiv1alpha1.Snapshot
+			foundSnapshotWithOpenedPR, err = a.findSnapshotWithOpenedPR(&applicationComponent, a.pipelineRun.Namespace, prGroupHash)
+			if err != nil {
+				return false, err
+			}
+			if foundSnapshotWithOpenedPR != nil {
+				a.logger.Info("Opened PR/MR in snapshot is found")
+				componentsWithOpenPRMR = append(componentsWithOpenPRMR, applicationComponent.Name)
+			}
+		}
+	}
+	if len(componentsWithOpenPRMR) < 2 {
+		a.logger.Info(fmt.Sprintf("The number %d of components affected by this PR group %s is less than 2, skipping group snapshot creation", len(joinedComponents), prGroup))
+		return false, nil
+	}
+	return true, nil
+}
+
+// findSnapshotWithOpenedPR is copied from snapshot_adapter.go and it find the snapshot with opened PR/MR for the given component
+func (a *Adapter) findSnapshotWithOpenedPR(applicationComponent *applicationapiv1alpha1.Component, nameSpace, prGroupHash string) (*applicationapiv1alpha1.Snapshot, error) {
```

Comment:
Not sure how complex it would be to extract both of these copied functions out of their adapters and into the `gitops` package to prevent code duplication, so @hongweiliu17  I leave it up to your judgement if that effort would be worth it.

Reply:
I thought to combine them together, let me rethink it. thanks for all comments and I will update them.

Reply:
Moved to status package. Thanks.

Reply:
Thanks, nicely done!